PLT-2019: Add functionality to package

timothestes opened the pull request:
### Summary

This pull request adds functionality to authenticate to AWS CodeArtifact repositories before running `poetry add` or `poetry install` commands. The plugin checks for AWS CodeArtifact sources in the `pyproject.toml` file and sets the necessary environment variables to authenticate using the specified AWS profile.

### What was changed and why it was changed

**README.md**
- Expanded the README to include instructions for configuring the plugin, how it works under the hood, installation and uninstallation steps, and dependencies.

**makefile**
- Added a `makefile` to simplify the setup process with `poetry install` and running tests with `pytest`.

**poetry.lock**
- Generated a new `poetry.lock` file to include the new dependencies.

**poetry_partifact_plugin/plugin.py**
- Created the main plugin file, which includes methods for reading and parsing the `pyproject.toml` file, checking for CodeArtifact sources, getting the AWS profile name, and setting the necessary environment variables for authentication.

**pyproject.toml**
- Added a new `pyproject.toml` file that defines the package, dependencies, build system, and plugin configuration.

**tests/test_plugin.py**
- Added a new test file to include test cases for the plugin.

**.github/actions/setup-python-poetry/action.yaml**
- This file defines a new GitHub Action called "Install Python and Poetry" which installs the specified Python and Poetry versions and provides options for installing Python dependencies or only development dependencies.

**.github/workflows/publish-package.yaml**
- This new workflow is designed to automatically publish a Python package upon pushing to the main branch if the `pyproject.toml` file has been changed. The workflow first checks if the version has been changed, sets the Python version, sets up Python and Poetry, builds the package, and then publishes it to PyPI (this step is currently commented out and awaiting the creation of an official Amino organization on PyPI).

**.github/workflows/unit-tests.yaml**
- This new workflow is designed to run unit tests on every pull request that touches `.py` files, when the workflow is manually triggered, or on a scheduled basis every 15 days. The workflow runs tests on multiple Python versions and platforms specified in a matrix. The workflow is divided into two jobs: `prepare` and `test`. The `prepare` job sets up Python and Poetry, builds the package distribution files, and stores them as artifacts. The `test` job then retrieves the pre-built distribution files, installs the package, and runs the tests.

timothestes committed:
feat: add plugin, tests, and action

fix license indentation

timothestes committed:
add License to readme

timothestes committed:
fix(GhActions): comment out publish step, add build conditions

timothestes committed:
tone down gitignore

timothestes committed:
switch to new setup-python-poetry action

cjduffett commented on .github/actions/setup-python-poetry/action.yaml:
```diff
@@ -0,0 +1,47 @@
+      uses: actions/setup-python@v4
+      with:
+        python-version: ${{ inputs.python-version }}
+        cache: "poetry"
```
What does this do? Perhaps should add a comment here to indicate what's being cached, and why?

timothestes replied:
The `cache: "poetry"` option enables caching for Python dependencies that are managed by Poetry, which can help speed up build times instead of having to install everything each time.

Behind the scenes, the action will create a hash based on the poetry.lock file, which contains the exact version of each dependency. This ensures that the cache is only used when the dependencies have not changed. When the workflow runs again, and the cache key matches the current poetry.lock file, the cached dependencies will be used instead of downloading and installing them from scratch.

I believe the cache is stored somewhere in Github and lives in the repository (not sure of the details about cache storage)

cjduffett commented on .github/actions/setup-python-poetry/action.yaml:
```diff
@@ -0,0 +1,47 @@
+      run: poetry install --only dev
+      shell: bash
+
+    - name: Install All Dependencies
```
nit: change "All Dependencies" to "Project Dependencies"

cjduffett commented on .github/workflows/publish-package.yaml:
```diff
@@ -0,0 +1,61 @@
+    paths: # assume to only publish package when pyproject.toml file changes
+      - pyproject.toml
+  
+env: # default versions
+  POETRY_VERSION: 1.3.1
+  PYTHON_VERSION: 3.11.2
```
Ideally both these env vars, and the values used in the "Install Poetry" step above, can come from files in the repository: we typically maintain `.python-version` and `.poetry-version` files for this reason. Is it possible to source these config options from the same project file in all cases?

timothestes replied:
I've already added a step that attempts to read the .python-version file (if it exists). If there is a .python-version file, the default PYTHON_VERSION variable will be overwritten.

I can add the same logic for setting the POETRY_VERSION.

I'm hesitant to soley rely on .python-version and .poetry-version files b/c I know some repos don't have .poetry-version files.

cjduffett commented on .github/workflows/unit-tests.yaml:
```diff
@@ -0,0 +1,105 @@
+  pull_request: # Run in every PR that touches .py files
+    paths:
+      - '**.py*'
+  workflow_dispatch: # Allow manually triggering the workflow
```
This appears to be a key without a value?

timothestes replied:
Yes, although it appears weird, this indeed is a valid way to configure the workflow_dispatch option. See https://stackoverflow.com/a/62768214

cjduffett commented on .github/actions/setup-python-poetry/action.yaml:
```diff
@@ -0,0 +1,47 @@
+description: "Add Poetry, dependency manager for Python"
+
+inputs:
+  python-version: # Dynamically set Python version
```
It appears the GitHub Action naming convention is `kebab-case` instead of `snake_case`? Ideally I would expect variables we define to be snake_case like `python_version`, `poetry_version`, etc.

timothestes replied:
Yes, the github action naming convention is kebab case. If I changed this to be `snake_case`, it would clash will all the other github actions out there (and then we start having to do a mental guessing game: is it `snake_case` or is it `kebab-case`? Going to stick with `kebab-case` for now.

cjduffett commented on .github/workflows/unit-tests.yaml:
```diff
@@ -0,0 +1,105 @@
+          # - "3.9"
+        platform: # specify all the machines you want to test on
+          - ubuntu-latest
+          # - macos-latest
+          # - windows-latest
```
Should just remove these if not used?

cjduffett commented on .github/workflows/unit-tests.yaml:
```diff
@@ -0,0 +1,105 @@
+            only-dev-deps: true
+
+      - name: Build package distribution files
+        run: |
+          poetry build
```
nit: compact commands like `poetry build` or `make test` (below) can be on the same line as `run`?

cjduffett commented on poetry_partifact_plugin/plugin.py:
```diff
@@ -0,0 +1,108 @@
+import os
```
nit: Include module and class/function docstrings

cjduffett commented on poetry_partifact_plugin/plugin.py:
```diff
@@ -0,0 +1,108 @@
+CONFIG_PATH = "./pyproject.toml"
+
+
+class PartifactPlugin(ApplicationPlugin):  # type: ignore
```
What's the type error here?

timothestes replied:
This shouldn't be here (it was leftover from some code I was using as a reference for building a poetry plugin class)

cjduffett commented on poetry_partifact_plugin/plugin.py:
```diff
@@ -0,0 +1,108 @@
+
+
+class PartifactPlugin(ApplicationPlugin):  # type: ignore
+    name = "PartifactPlugin plugin"
```
nit: redundant "plugin"?

timothestes replied:
yep!

cjduffett commented on poetry_partifact_plugin/plugin.py:
```diff
@@ -0,0 +1,108 @@
+
+    def _get_sources(self, parsed_toml: dict) -> list:
+        """Return a list of all tool.poetry.source's in the parsed toml."""
+        return parsed_toml.get("tool", {}).get("poetry", {}).get("source", [])
```
Should this fail loudly if `tool.poetry.source` is not configured? I think it's preferable to raise a KeyError if required config is missing.

timothestes replied:
No, this function needs to run in my `self._pyproject_toml_has_codeartifact()` function which checks the pyproject.toml file for .codeartifact. in the source block. 

I don't want this to fail loudly because the desired behavior is that the function will simply do nothing if the `tool.poetry.source` block isn't found.

cjduffett commented on poetry_partifact_plugin/plugin.py:
```diff
@@ -0,0 +1,108 @@
+
+            cleo_io.write_line(
+                f"<fg=green>{self.name} successfully configured AWS CodeArtifact for {profile_name}</info>"
+            )
```
I would refactor this into another private function, to minimize the amount of code in the `try` block. When writing `try/except` statements, ideally only **one** line of code is executed in the `try` block so it's clear exactly which statement raised the exception.

cjduffett commented on pyproject.toml:
```diff
@@ -0,0 +1,37 @@
+name = "poetry-partifact-plugin"
+version = "0.1.0"
+description = "A poetry plugin that configures AWS CodeArtifact"
+authors = ["AminoEngineers <eng@amino.com>"]
```
nit: Change "AminoEngineers" to "Amino Engineering Team"

cjduffett commented on pyproject.toml:
```diff
@@ -0,0 +1,37 @@
+[tool.poetry.plugins."poetry.application.plugin"]
+poetry-partifact-plugin = "poetry_partifact_plugin.plugin:PartifactPlugin"
+
+[virtualenvs]
+create = true
+in-project = true
```
neat! In the past I've relied on a `poetry.toml` file to configure these options.

timothestes replied:
I actually don't think the way I declared it is valid. Switching to poetry.toml file...

cjduffett commented on tests/test_plugin.py:
```diff
@@ -0,0 +1,90 @@
+from poetry_partifact_plugin.plugin import PartifactPlugin
+
+
+@pytest.fixture
```
nit: Put these in a separate `conftest.py` file so it's easier to grok the test cases in this file?

timothestes committed:
address code review

add conditional statements for setting python and poetry versions

move fixtures to conftest

switch to poetry.toml and update authors

move try-catch code to separate function

add better class docstrings

add module docstring

remove bad comments

add cache comment

nit: change name of install step

nit: remove unused testing options

compact small commands

cjduffett commented on poetry_partifact_plugin/plugin.py:
```diff
@@ -0,0 +1,187 @@
+                f"<error>{self.name} failed to configure AWS CodeArtifact: \n\t{error}</>"
+            )
+
+    def _set_env_vars(self, parsed_toml: str) -> None:
+        """
+        Sets environment variables required for AWS CodeArtifact authentication.
+
+        Args:
+            parsed_toml: The parsed TOML dictionary.
+            
+        Raises:
+            Exception: If there's an error getting the profile name.
+        """
+        profile_name = self._get_profile_name(parsed_toml)
+        formatted_profile_name = profile_name.upper().replace("-", "_")
+        config = Configuration.load(profile_name, profile=profile_name)
+
+        # setting these env variables will allow poetry to connect to codeartifact
+        # https://python-poetry.org/docs/configuration/#using-environment-variables
+        os.environ[f"POETRY_HTTP_BASIC_{formatted_profile_name}_PASSWORD"] = get_token(
+            config
+        )
+        os.environ[f"POETRY_HTTP_BASIC_{formatted_profile_name}_USERNAME"] = "aws"
```
Not seeing a unit test for this - feels important since we want to make sure the correct env vars are set with the correct values. This will require some mocking and patching, for example mocking the return value of `get_token()` with a fake value.

cjduffett commented on tests/test_plugin.py:
```diff
@@ -0,0 +1,43 @@
+
+
+def test_object_initialization():
+    """Unsure of the best way to test a poetry plugin..."""
```
nit: Test docstrings typically describe what condition the test is evaluating. Often in the form of a "Should ..." statement. For example, this could read:
```py
"""Should correctly initialize the poetry plugin class."""
```

timothestes committed:
add unit tests for env_vars methods

timothestes committed:
add another assertion
